tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/aux/CherryPick.scala

```diff
@@ -573,29 +574,54 @@ class CherryPick(rewriter: SymbStateRewriter) {
-   * as relations, the implementation became trivial
+   * Picks a function from a set.
    *
    * @param funType a cell type to assign to the picked cell.
-   * @param oracle  a variable that stores which element (by index) should be picked, can be unrestricted
-   * @param funs    a sequence of cells that store functions
-   * @param state   a symbolic state
+   * @param oracle  a variable that stores which element (by index) should be picked, can be unrestricted.
+   * @param funs    a sequence of cells that store functions.
+   * @param state   a symbolic state.
    * @return a new symbolic state with the expression holding a fresh cell that stores the picked element.
    */
   def pickFun(funType: FunT, state: SymbState, oracle: Oracle, funs: Seq[ArenaCell], elseAssert: TlaEx): SymbState = {
     rewriter.solverContext.log("; CHERRY-PICK %s FROM [%s] {".format(funType, funs.map(_.toString).mkString(", ")))
     var nextState = state
-    // pick the relation
-    val relationT = FinSetT(TupleT(Seq(funType.argType, funType.resultType)))
-    nextState = pickSet(relationT, nextState, oracle, funs map state.arena.getCdm, elseAssert)
-    val pickedRelation = nextState.asCell
-    // create a fresh cell to hold the function
-    nextState = nextState.setArena(nextState.arena.appendCell(funType))
-    val funCell = nextState.arena.topCell
-    val newArena = nextState.arena.setCdm(funCell, pickedRelation)
-    rewriter.solverContext.log(s"; } CHERRY-PICK $funCell:$funType")
-    // That's it! Compare to pickFunPreWarp.
-    nextState.setArena(newArena).setRex(funCell.toNameEx)
+    rewriter.solverContext.config.smtEncoding match {
+      case `arraysEncoding` =>
+        // Here we have a simplified version of pickBasic, using SMT equality instead of lazy equality.
+        // It is so because in the arrays encoding we pick a function directly, instead of picking a set of relations.
+        rewriter.solverContext
+          .log("; CHERRY-PICK %s FROM [%s] {".format(funType, funs.map(_.toString).mkString(", ")))
+        nextState = nextState.updateArena(_.appendCell(funType))
+        val funCell = nextState.arena.topCell
+        val asserts = funs map { el => OperEx(TlaOper.eq, funCell.toNameEx, el.toNameEx) }
+        rewriter.solverContext.assertGroundExpr(oracle.caseAssertions(nextState, asserts :+ elseAssert))
+
+        // The cell with the picked function has to have its associated metadata, so we query the solver for the model
+        // produced, take the picked index, and update its associated function with the correct domain and cdm.
+        rewriter.solverContext.sat()
+        val index = oracle.evalPosition(rewriter.solverContext, nextState)
+        nextState = nextState.updateArena(_.setDom(funCell, nextState.arena.getDom(funs(index))))
+        nextState = nextState.updateArena(_.setCdm(funCell, nextState.arena.getCdm(funs(index))))
```

Comment:
This code accesses an SMT model, which may be unavailable. Moreover, I am not sure how sound it is. I see that you need this to later decode the model. Let's think how to do it without querying the solver.

Maybe we could introduce another cell that statically stores that a choice should be done.

Reply:
The problem is that since we delegate the picking to the solver with `oracle.caseAssertions` (on the line right above the highlighted extract) we cannot know which cell was picked without getting the model.

> Maybe we could introduce another cell that statically stores that a choice should be done.

We might need to completely rework this method. I'm open to suggestions, but I don't get what do you mean here.

Reply:
`pickFun` was reworked after a discussion with @konnov. It no longer calls `solverContext.sat()`.